### DISCUSSION TITLE
what does the set method signature look like?

### DISCUSSION THREAD

**Discussion by budarin** [pmndrs/zustand#1867] [Q&A] [answered]
Interested in the question: can `set` return anything other than `void`?

Now, in case of an error inside the method, I have to throw an exception
I want to avoid throwing exceptions in the methods of the store and use the technique of returning the object `{ result, error }`

I also want not to litter the description of the store with texts of methods (there may be many of them)- I want to describe methods in separate modules and import them into the store , but the description of the `set` method type is not appropriate

**Comment by dai-shi** [comment, marked as answer]
By method, do you mean a function you pass to `set`?
I think throwing is correct and expected. You could wrap `set` and create the version that returns `{ result, error }` on your end.

**Reply by budarin** [reply]
throwing an exception is right but is slow and hard to process
could you give me an example of the code ?

```js
addTodo: (todo: Todo) => {
    return set((state) => {
       ...
    });
}
```
By method, I mean - `addTodo` 
I need to have the interface:
```ts
addTodo(todo) => { result } | { error }
```

or it's impossible?

**Reply by dai-shi** [reply]
I think what you want is this.
```js
const useTodoStore = create((set, get) => ({
  ...
  addTodo: (todo) => {
    const prevState = get();
    const nextState = ...;
    if (...) {
      set(nextState);
    }
    return { anything... };
  }
}));
```

**Reply by budarin** [reply]
the 2nd part of the question remains: how do I put the method description in an external module?

there is no way to do without a description of the `set` and `get` types, and I do not know their signatures

**Reply by dai-shi** [reply]
That might be tough. I'd try this: `typeof useTodoStore.setState` and `typeof useTodoStore.getState`.

> how do I put the method description in an external module?

I'm not sure if I understand the goal though.

**Reply by budarin** [reply]
if you describe all the methods in the store, the module may turn out to be huge and it will not be convenient to read it
So I want to have the description `addTodo` in it's own module and import it into TodoStore's module

```ts
// addTodo.ts
export function addTodo(todo) {
   return (get:?, set: ?) => {
    ...
  }
}
```

```ts
// TodoStore
import { addTodo } from './addTodo';

const useTodoStore = create((set, get) => ({
  ...
  addTodo: addTodo,
  ...
})
```

**Reply by dai-shi** [reply]
Thanks for the clarification! Can you try `typeof useTodoStore.setState`? It's kind of circular, so not sure if it works.

**Reply by budarin** [reply]
Thanks, I'll try tomorrow and will let you know about
I'm afraid that there will be cyclic connections of modules ...

**Reply by dai-shi** [reply]
Yeah. If it's just a TypeScript issue, a small reproduction with https://www.typescriptlang.org/play/ would be helpful.

**Reply by budarin** [reply]
Everything is OK!
simply imported `useTodoStore` and used `setState` and `getSate`

```ts
// addTodo.ts
import { useTodoStore } from './useTodoStore'

export function addTodo(todo) {
    const prevState = useTodoStore.getState();
    const nextState = ...;
    if (...) {
      useTodoStore.setState(nextState);
    }
    return { anything... };
```

**Reply by dai-shi** [reply]
Yeah, that works too.
I'm kind of pushing such a style as newer best practice: https://github.com/pmndrs/zustand/blob/a53433543cf36f6379242f3212363eb770340dd1/docs/guides/practice-with-no-store-actions.md
